core/service/src/cryptography/encryption.rs

```diff
@@ -0,0 +1,641 @@
+                Ok(UnifiedPublicEncKey::MlKem1024(inner))
+            }
+            // in case the old deserialization fails, try the new format
+            Err(_) => tfhe::safe_serialization::safe_deserialize::<UnifiedPublicEncKey>(
+                std::io::Cursor::new(&bytes),
+                crate::consts::SAFE_SER_SIZE_LIMIT,
+            )
+            .map_err(|e| CryptographyError::DeserializationError(e.to_string())),
+        }
+    }
+}
+
+impl UnifiedPublicEncKey {
+    /// Expect the inner type to be the default MlKem512 and return it, otherwise panic
+    pub fn unwrap_ml_kem_512(self) -> PublicEncKey<ml_kem::MlKem512> {
+        match self {
+            UnifiedPublicEncKey::MlKem512(pk) => pk,
+            _ => panic!("Expected MlKem512 public encryption key"),
+        }
+    }
+}
+
+// Alias wrapping the ephemeral public encryption key the user's wallet constructs and the server
+// uses to encrypt its payload
+// The only reason this format is not private is that it is needed to handle the legacy case, as we do this by distinguishing between 512 and 1024 bit keys
+pub struct PublicEncKey<C: KemCore>(pub(crate) C::EncapsulationKey);
+
+impl<C: KemCore> Eq for PublicEncKey<C> {}
+impl<C: KemCore> PartialEq for PublicEncKey<C> {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.as_bytes().as_bytes() == other.0.as_bytes().as_bytes()
+    }
+}
+
+impl<C: KemCore> Serialize for PublicEncKey<C> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_bytes(&self.0.as_bytes())
+    }
+}
+
+impl<C: KemCore> Named for PublicEncKey<C> {
+    const NAME: &'static str = "PublicEncKey";
+}
+
+/// workaround because clone doesn't get derived for this type
+impl<C: KemCore> Clone for PublicEncKey<C> {
+    fn clone(&self) -> Self {
+        let buf = self.0.as_bytes();
+        PublicEncKey(C::EncapsulationKey::from_bytes(&buf))
+    }
+}
+
+impl<C: KemCore> std::fmt::Debug for PublicEncKey<C> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PublicEncKey")
+            .field("encapsulation_key", &self.0.as_bytes())
+            .finish()
+    }
+}
+
+impl<C: KemCore> tfhe_versionable::Versionize for PublicEncKey<C> {
+    type Versioned<'vers>
+        = &'vers PublicEncKey<C>
+    where
+        C: 'vers;
+
+    fn versionize(&self) -> Self::Versioned<'_> {
+        self
+    }
+}
+
+impl<C: KemCore> tfhe_versionable::VersionizeOwned for PublicEncKey<C> {
+    type VersionedOwned = PublicEncKey<C>;
+    fn versionize_owned(self) -> Self::VersionedOwned {
+        self
+    }
+}
+
+impl<C: KemCore> tfhe_versionable::Unversionize for PublicEncKey<C> {
+    fn unversionize(
+        versioned: Self::VersionedOwned,
+    ) -> Result<Self, tfhe_versionable::UnversionizeError> {
+        Ok(versioned)
+    }
+}
+
+impl<C: KemCore> tfhe_versionable::NotVersioned for PublicEncKey<C> {}
```

Comment:
I forgot if it was like this before, but this means this type is not versioned. is this what we want?

Reply:
I am not sure. I thought it was still versioned, since that trait is implemented above. We definitely want the type versioned! Since it may change in the future. But I can easily have missed something. I thought that the current trait implementations would allow us to do manual updates later on.
Do you have any suggestions how to ensure versioning?

Reply:
I am not sure `impl_generic_versionize` can be used because of the generics. But I removed the `NotVersioned` implementation in d50dab5631013555042c91b8635fd2028e621413

Reply:
So I think this relates a lot to the discussion [here](https://zama-ai.slack.com/archives/C05UD8KC0UD/p1762247938385819) as well. 
I would argue that we need a more flexible signature implementation to be future proof (as covered in [this issue](https://github.com/zama-ai/kms-internal/issues/2781)). In that situation we would need to wrap `PublicEncKey` in an enum that would contain any future versions of signing keys we might need. 
I am not sure what the best path is at this point. Maybe it is to keep `NotVersioned` as we use `PublicEncKey` verbatim as an ecdsa key?

Reply:
Added back in d737f1c2b4ed46ca6b0799dd69dbf56884ef2a3d